Added gamemodes. New mode: Fast.

J4RV opened the pull request:
Also added Codacy and removed some code smells.

J4RV commented:
Crap, I wanted to merge this with my master branch. 

Check it out if you want. The fast mode means that every player has 60 seconds to play their turn. If they don't, the bot automatically skips them.

jh0ker commented on game.py:
```diff
@@ -76,6 +88,16 @@ def turn(self):
+        if not self.deck.cards:
+            self.set_mode("classic")
+
+        while not self.last_card or self.last_card.special:
```
The reason I had the `self.deck = Deck()` in this loop was because you potentially remove special cards from the deck completely. You could add something like `if self.last_card.special: self.deck.dismiss(self.last_card)`

J4RV replied:
I don't think I get the potential to remove special cards from the deck with this code, can you explain the problem a bit more?

jh0ker replied:
This is a loop that draws cards until a condition is met. What happens to the cards that are drawn?

jh0ker commented on game.py:
```diff
@@ -106,6 +128,7 @@ def play_card(self, card):
             self.turn()
         else:
             self.logger.debug("Choosing Color...")
+            self.logger.info("Played special card with hand: {hand}".format(hand=self.current_player.cards))
```
Do we need this log?

J4RV replied:
Nope, not really. That's there because I have a friend that said "hey, I didnt bluff!" a couple times, with this I can prove him wrong. I'll remove it for the next pull request.

jh0ker commented on actions.py:
```diff
@@ -0,0 +1,220 @@
+from user_setting import UserSetting
+from utils import send_async, display_name, game_is_running
+
+logging.basicConfig(
```
There's another `logging.basicConfig` in `bot.py`, I think this one should be removed

J4RV replied:
I did it this way because importing logger from bot caused a circular dependency. I'll move logger to shared_vars and use just one then.

jh0ker replied:
I wasn't referring to the `logger` though

jh0ker commented on actions.py:
```diff
@@ -0,0 +1,220 @@
+        self.job_queue = job_queue
+
+
+# TODO do_skip() could get executed in another thread (it can be a job), so it looks like it can't use game.translate?
```
Do you mean the `_` or `__` function fails? Can you post the error that happens?

J4RV replied:
Just did a quick test, adding a __ function inside do_skip and letting it trigger in a fast game:
https://i.gyazo.com/3b95384bfdf04d4d62ce6b44e1430f6c.png

Here's the text trace:
2017-10-23 19:08:17,163 - utils - ERROR - Message text is empty
Traceback (most recent call last):
  File "F:\PROJECTS\mau_mau_bot\utils.py", line 92, in send_async
    bot.sendMessage(*args, **kwargs)
  File "C:\Users\Jarv\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telegram\bot.py", line 57, in decorator
    result = func(self, *args, **kwargs)
  File "C:\Users\Jarv\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telegram\bot.py", line 69, in decorator
    return self._message_wrapper(url, data, *args, **kwargs)
  File "C:\Users\Jarv\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telegram\bot.py", line 165, in _message_wrapper
    result = self._request.post(url, data, timeout=kwargs.get('timeout'))
  File "C:\Users\Jarv\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telegram\utils\request.py", line 271, in post
    **urlopen_kwargs)
  File "C:\Users\Jarv\AppData\Local\Programs\Python\Python36-32\lib\site-packages\telegram\utils\request.py", line 209, in _request_wrapper
    raise BadRequest(message)
telegram.error.BadRequest: Message text is empty

jh0ker commented on actions.py:
```diff
@@ -0,0 +1,220 @@
+        try:
+            gm.leave_game(skipped_player.user, chat)
+            send_async(bot, chat.id,
+                       text="{name1} was skipped four times in a row "
```
Seems like "four times in a row" is not always the case anymore, right?

jh0ker commented on actions.py:
```diff
@@ -0,0 +1,220 @@
+
+        except NotEnoughPlayersError:
+            send_async(bot, chat.id,
+                       text="{name} was skipped four times in a row "
```
See above

jh0ker commented on bot.py:
```diff
@@ -19,41 +19,39 @@
     format='%(asctime)s - %(name)s - %(levelname)s - %(message)s',
-    level=logging.INFO)
+    level=logging.INFO
+    ,filename='log.log'
```
I would prefer logging to stdout - if one wants to log to a file they can use the `>` operator

jh0ker commented on bot.py:
```diff
@@ -292,14 +298,17 @@ def start_game(bot, update, args):
-        elif len(game.players) < 2:
+        elif len(game.players) < MIN_PLAYERS:
             send_async(bot, chat.id,
                        text=_("At least two players must /join the game "
```
again, it's not "at least two" anymore

jh0ker commented on bot.py:
```diff
@@ -637,134 +626,30 @@ def process_result(bot, update):
+        player.waiting_time = WAITING_TIME
         send_async(bot, chat.id,
-                   text=__("Waiting time for {name} has been reset to 90 "
+                   text=__("Waiting time for {name} has been reset to {time} "
```
You're changing the translation string here... did you account for that in `unobot.pot` and locales? Just to make sure it doesn't break completely

J4RV replied:
I don't know how unobot.pot works, but I have been playing in spanish just fine. The translation on the changed messages would have to be changed, yeah.

jh0ker commented on bot_config.py:
```diff
@@ -0,0 +1,4 @@
+DEFAULT_GAMEMODE = "fast"
```
Can you put a comment listing the different options here?

jh0ker replied:
And also, are we using `bot_config.py` or `config.json`?

jh0ker commented on game.py:
```diff
@@ -33,6 +33,8 @@ class Game(object):
     draw_counter = 0
     players_won = 0
     starter = None
+    mode = 'classic'
```
Shouldn't this be set from the config?

jh0ker commented on game.py:
```diff
@@ -76,6 +88,16 @@ def turn(self):
+    def _first_card_(self):
+        # In case that the player did not select a game mode
+        if not self.deck.cards:
+            self.set_mode("classic")
```
again, shouldn't this be set from the config?

jh0ker replied:
Also, what does setting the mode help if there are no cards in the deck?

jh0ker commented:
@J4RV I posted a code review! Thank you already for your amazing help :)

J4RV commented:
Great, I'll try to fix these things next week!

J4RV commented:
Fixed most of them, except:

The __ function still fails in do_skip if it's called from a thread (on fast mode)

Kept gameplay options in bot_config.py. Maybe changing its name to gameplay_config or something like that is a good idea.

Fixed translation strings in spanish (skipped 4 times and min players). It doesnt break in other languages, but shows the english translation.

jh0ker commented:
I left two more comments on the `_first_card_` method, and I think you misunderstood what I meant with the logging.

Yes, you should use one `logger` per file, but you should only do `logging.basicConfig` once.

J4RV commented:
Got it this time! I added a couple comments to that section to explain it.

jh0ker commented:
Thanks :+1: I'll do another full review tomorrow hopefully, with all these changes I want to be thorough. As for the translation threading issue - I guess we have to live with that for the time being, and perhaps address this in a future PR. I don't want to add more complexity to this PR. Oh and BTW I would appreciate it if you renamed `bot_config.py` to `gameplay_config.py` or something similar, like you suggested. And document this in the README as well.

J4RV commented:
Excellent! I'll do it right now 👌

DO97 commented:
When will it be available?

DO97 commented:
Up

J4RV commented:
Hey @jh0ker, did you check the code? Do you think it needs any changes?

jh0ker commented:
let's see :P
